File: leetcode/Ruby/No3.longest_substring_without_repeating_characters.rb
=begin
Difficulty:
Medium

Desc:
Given a string, find the length of the longest substring without repeating characters.

Examples:
Given "abcabcbb", the answer is "abc", which the length is 3.
Given "bbbbb", the answer is "b", with the length of 1.
Given "pwwkew", the answer is "wke", with the length of 3.
Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
=end

# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
  h = Hash.new
  max = 0
  len = 0
  s.chars.each_with_index do |char, index|
    if h[char] != nil then
      max = [max, len].max
      i = h[char]
      len = index - i
      i += 1
      h = Hash.new
      while i < index do
        h[s[i]] = i
        i += 1
      end
    else
      len += 1
    end
    h[char] = index
  end
  [max, len].max
end

def length_of_longest_substring2(s)
  h = Hash.new
  start = 0
  i = 0
  len = 0
  while i < s.size do
    char = s[i]
    if h.has_key?(char) then
      start = [start, h[char] + 1].max
    end
    len = [len, i + 1 - start].max
    h[char] = i
    i += 1
  end
  return len
end

# Test case
puts length_of_longest_substring2("abcabcbb") # 3
puts length_of_longest_substring2("pwwkew") # 3
puts length_of_longest_substring2("bbbbb") # 1
puts length_of_longest_substring2("bbabbb") # 2

File: leetcode/C#/No3.longest-substring-without-repeating-characters.cs
/*
 * Difficulty:
 * Medium
 *
 * Desc:
 * Given a string, find the length of the longest substring without repeating characters.
 *
 * Example 1:
 * Input: "abcabcbb"
 * Output: 3 
 * Explanation: The answer is "abc", with the length of 3. 
 *
 * Example 2:
 * Input: "bbbbb"
 * Output: 1
 * Explanation: The answer is "b", with the length of 1.
 *
 * Example 3:
 * Input: "pwwkew"
 * Output: 3
 * Explanation:
 * The answer is "wke", with the length of 3. 
 * Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
*/

public class Solution {
    public int LengthOfLongestSubstring(string s) {
        int res = 0;
        int i = 0;
        int j = 0;
        Dictionary<char, int> dict = new Dictionary<char, int>();

        while (j < s.Length) {
            char str = s[j];

            if (dict.ContainsKey(str)) {
                res = Math.Max(res, j - i);

                int k = dict[str];
                while (i <= k) {
                    dict.Remove(s[i]);
                    i += 1;
                }
            }

            dict[str] = j;
            j += 1;
        }

        return Math.Max(res, j - i);
    }
}